velox/functions/prestosql/aggregates/ArrayAggAggregate.cpp

```diff
@@ -97,26 +113,74 @@ class ArrayAggAggregate : public exec::Aggregate {
     auto vector = (*result)->as<ArrayVector>();
     VELOX_CHECK(vector);
     vector->resize(numGroups);
+    auto* resultOffsets =
+        vector->mutableOffsets(numGroups)->asMutable<vector_size_t>();
+    auto* resultSizes =
+        vector->mutableSizes(numGroups)->asMutable<vector_size_t>();
+    auto& elements = vector->elements();
+    vector_size_t arrayOffset = 0;
+
+    if (useClusteredAccumulator()) {
+      vector_size_t numTotalElements = 0;
+      for (int32_t i = 0; i < numGroups; ++i) {
+        auto* accumulator = value<ClusteredAccumulator>(groups[i]);
+        for (auto& source : accumulator->sources) {
+          numTotalElements += source.size;
+        }
+      }
+      elements->resize(numTotalElements);
+      VectorPtr* currentSource = nullptr;
+      BaseVector::CopyRange range;
+      range.sourceIndex = -1;
+      range.targetIndex = 0;
+      range.count = std::numeric_limits<vector_size_t>::min();
+      for (int32_t i = 0; i < numGroups; ++i) {
+        auto* accumulator = value<ClusteredAccumulator>(groups[i]);
+        resultOffsets[i] = range.targetIndex;
+        vector_size_t arraySize = 0;
+        for (auto& source : accumulator->sources) {
+          arraySize += source.size;
+          if (!currentSource) {
+            currentSource = &source.vector;
+            range.sourceIndex = source.offset;
+            range.count = source.size;
+          } else if (currentSource->get() != source.vector.get()) {
+            elements->copyRanges(currentSource->get(), folly::Range(&range, 1));
+            currentSource = &source.vector;
+            range.sourceIndex = source.offset;
+            range.targetIndex += range.count;
+            range.count = source.size;
+          } else {
+            VELOX_DCHECK_EQ(range.sourceIndex + range.count, source.offset);
+            range.count += source.size;
+          }
+        }
+        resultOffsets[i] = arrayOffset;
+        resultSizes[i] = arraySize;
+        arrayOffset += arraySize;
+      }
+      if (currentSource) {
+        elements->copyRanges(currentSource->get(), folly::Range(&range, 1));
+      }
 
-    auto elements = vector->elements();
-    elements->resize(countElements(groups, numGroups));
-
-    uint64_t* rawNulls = getRawNulls(vector);
-    vector_size_t offset = 0;
-    for (int32_t i = 0; i < numGroups; ++i) {
-      auto& values = value<ArrayAccumulator>(groups[i])->elements;
-      auto arraySize = values.size();
-      if (arraySize) {
-        clearNull(rawNulls, i);
-
-        ValueListReader reader(values);
-        for (auto index = 0; index < arraySize; ++index) {
-          reader.next(*elements, offset + index);
+    } else {
+      elements->resize(countElements(groups, numGroups));
```

Comment:
nit: Would it make sense to enhance countElements to consider useClusteredAccumulator() and reuse it in both branches (or even outside of the if-else block)?